hopr/api/src/ct/traits.rs

```diff
@@ -0,0 +1,36 @@
+use multiaddr::PeerId;
+
+pub use hopr_network_types::types::DestinationRouting;
+
+use super::{PathTelemetry, Telemetry};
+
+#[derive(thiserror::Error, Debug)]
+pub enum TrafficGenerationError {
+    #[error("timed out for near neighbor probe '{0:?}'")]
+    ProbeNeighborTimeout(PeerId),
+
+    #[error("timed out for loopback probe")]
+    ProbeLoopbackTimeout(PathTelemetry),
+}
+
+/// A trait for types that can produce a stream of cover traffic routes.
+///
+/// The basic assumption is that the implementor will provide the logic
+/// to choose suitable route candidates for cover traffic based on a
+/// custom algorithm.
+///
+/// The implementor should ensure that the produced routes are indefinite,
+/// since the exhaustion of the stream might result in termination of the
+/// cover traffic generation.
+pub trait TrafficGeneration {
+    fn build(
+        self,
+    ) -> (
+        impl futures::Stream<Item = DestinationRouting> + Send,
+        impl futures::Sink<std::result::Result<Telemetry, TrafficGenerationError>, Error = impl std::error::Error>
+        + Send
+        + Sync
+        + Clone
+        + 'static,
+    );
+}
```

Comment:
Consider adding `type Error: std::error::Error + Send + Sync + 'static` as the trait type, so that the implementation has the freedom defining its implementation-specific error states. Here should be only such errors that carry an immediate value to the caller and are a part of the trait's contract.

Check the other APIs.

Reply:
This cannot be. The API user needs to know and distinguish timeouts from other types of errors.

Reply:
Removed, as probing is now a stream decorator.